### DISCUSSION TITLE
LDM Notes for February 26th, 2024

### DISCUSSION THREAD

**Discussion by 333fred** [dotnet/csharplang#7971] [LDM Notes]
https://github.com/dotnet/csharplang/blob/main/meetings/2024/LDM-2024-02-26.md

## Agenda

- `ref struct`s in generics
- Collection expressions

**Comment by TahirAhmadov** [comment]
For the WinForms etc collections, I vote option 2 - very strange to modify the language because of a few very old types, which need to be updated anyway.

**Reply by KennethHoff** [reply]
If the only reason for changing the language is to "appease" WinForms and/or WPF then that doesn't make sense. That's their problem and it's up to them to fix it.

If however there's a more general underlying problem then that's a different story, but - at least according to the provided notes - that doesn't seem to be the case. I vote for option 2 as well. This can always change in the future, as this seemingly is not a backwards-compatability issue. 2.2 feels to me the "best" solution - one that should've happened years ago - but it brings its own set of problems, so 2.1 is the easiest.

**Reply by CyrusNajmabadi** [reply]
> If the only reason for changing the language is to "appease" WinForms and/or WPF then that doesn't make sense. That's their problem and it's up to them to fix it.

It's not their problem.  They designed their types in an expected fashion based in a way that has worked fine for 20+ years.  The patterns they use completely work with how we designed collection-initializers back in 3.0.  Furthermore, many of these are simply not going to get updated *period*.  The risk of breakage on their part is enormous, and they have tons of customers who depend on this stuff working.

We are not jettison'ing core parts of our user base that have done nothing wrong at all (either hte library authors, or their consumers).  Such a move is incredibly customer hostile, and anathema to our stated compat goals.

**Reply by CyrusNajmabadi** [reply]
> very strange to modify the language because of a few very old types, which need to be updated anyway.

We're modifying the language because we took a break early in the year to have collection expressions no longer work as originally designed (where they supported all these collection initializer cases).  We *want* people who use collection initializer types to be able to move them forward.  These types are not odd.  They're defined using *exactly* the pattern we said would work and which we worked *with* WinForms/WPF to adopt.   And we very intentionally designed collection exprs originally such that these would move forward.

Making these not work, after decades of them properly implementing the pattern *we told them to implement* is puling the rug out from underneath them for no good reason.

**Reply by KennethHoff** [reply]
> If the only reason for changing the language is to "appease" WinForms and/or WPF then that doesn't make sense. That's their problem and it's up to them to fix it.
>> It's not their problem. They designed their types in an expected fashion based in a way that has worked fine for 20+ years. The patterns they use completely work with how we designed collection-initializers back in 3.0.

That tells me that it is _not_ to appease WinForms and/or WPF and there is a more underlying issue, in which case it goes against the premise in the notes:
> Finally, we looked at some of the consequences of our [previous decision](https://github.com/dotnet/csharplang/blob/main/meetings/2024/LDM-2024-01-10.md#collection-expressions-conversion-vs-construction) with how the impact a few real-world types in WinForms and WPF.

In which case I agree that point 2 is not ideal.

---

I have a question about point 5:

> Include explicit `Add` implementations in the search for the `Add(object)` method for the purposes of conversion.

The word "explicit"; what does that mean? Just a bog-standard method, explicit interface implementations, or something else? Does it mean as long as the collection type has an `Add(T)` method - regardless of whether it is an `IEnumerable<T>` - you can use a collection expression of type `T`?

If that's the case, then I think point 5 is quite intuitive.

**Reply by CyrusNajmabadi** [reply]
> That tells me that it is not to appease WinForms and/or WPF and there is a more underlying issue, in which case it goes against the premise in the notes:

The premise is accurate.  WinForms/WPF are just the immediate, very visible, consequences of that decision that we didn't consider when making it :)

It's not about appeasing them.  It's that they demonstrate that the problem is not hypothetical or niche.  But it's very real world and present in significant libraries that are in widespread use in the ecosystem :)

**Reply by CyrusNajmabadi** [reply]
> The word "explicit"; what does that mean? 

An explicitly implemented method.  So if something implements `IList` (the non-generic one) but their impl of `.Add(object)` is explicit (not implicit) we can find that as use that for doing our validity checks.

**Reply by KennethHoff** [reply]
(I realize the word "appease" might not have been the best choice of words)

**Reply by TahirAhmadov** [reply]
@CyrusNajmabadi let's take the `ColumnHeaderCollection` type specifically. Does implementing `IEnumerable<ColumnHeader>` allow it to be constructed using a collection expression? And if so, what's the harm (potential break) in implementing it?

On the other hand, given that a lot of these collections are initialized in the ".Designer.cs" files, option 2.1 is also fine - I write WinForms programs from time to time and I don't really care about these collections supporting collection expressions.

**Reply by CyrusNajmabadi** [reply]
> Does implementing IEnumerable<ColumnHeader> allow it to be constructed using a collection expression?

Probably.  Would still have to make sure that if the user provides a string value that that works (like it would for collection-initializers).

> And if so, what's the harm (potential break) in implementing it?

Making any sort of inheritance change carries high risk.  Especially in codebases that are literally decades old.  Types may now match checks that they didn't previously match, or go through codepaths they didn't go through before.  The bar for changing these systems is *enormous* because the risk is extremely high.

> I write WinForms programs from time to time and I don't really care about these collections supporting collection expressions.

I don't see how that's relevant. For the customers doing the normal expected way of manually creating UI frameworks, we would like collection-exprs to work.  It was a core design principle from the beginning that we were not intentionally leaving people behind unless we really saw no mechanism whatsoever to move them forward.  Furthermore, these scenarios *did* work when we shipped collection expressions.  On top of that, we *told* people they would work *and* our own features push to code in this fashion.  But a *post ship* change ended up breaking this.  We want to ensure that we go back to a state where these normal and expected cases continue to work (As we said they would work).

A core virtue of collection-exprs is that it tries to provide a unifying way to produce collections for *all* the myriad ways users were previously doing it before.  Taking such a core area (collection-initializer types) which have been the *primary* way of doing this for the ecosystem for decades, as well as reverting support *we already shipped* for core frameworks (like WinForms/WPF) doesn't benefit *anyone*.  It just bifurcates the ecosystem, which was a super pressing thing we did *not* want to do here.

**Reply by TahirAhmadov** [reply]
> Probably. Would still have to make sure that if the user provides a string value that that works (like it would for collection-initializers).

Do you mean the intention is to support something like `[columnHeader1, columnHeader2, "Column 3"]`, etc.? I think this looks weird and should not be supported.
Even `["Column 1", "Column 2"]` should not be supported, IMO. 
The only thing one can should be able to do is `[columnHeader1, columnHeader2, columnHeader3]`.

> Making any sort of inheritance change carries high risk.

I just cannot imagine a single snippet of code along the lines of:
```cs
object[] bagOfStuff = new object[10];
bagOfStuff[0] = listView.Columns;
bagOfStuff[1] = new List<ColumnHeader>();
// ...
var list = bagOfStuff.OfType<IEnumerable<ColumnHeader>>().First();
```
But then it's a weakness of mine to expect things to be logical (especially in my understanding of logic), so it's possible you're right about the risk.

> It just bifurcates the ecosystem, which was a super pressing thing we did not want to do here.

In general, I agree with this sentiment. But...

>  For the customers doing the normal expected way of manually creating UI frameworks, we would like collection-exprs to work.

For this specific type, though, I fail to see how C.E.s are helpful. The [ListView.Columns](https://learn.microsoft.com/en-us/dotnet/api/system.windows.forms.listview.columns?view=windowsdesktop-8.0) property is `get` only, and the backing field is `readonly`. From my reading of the code, the `ColumnHeaderCollection` class is just an "interface" and not really a "collection".

Overall, this looks like more of a case-by-case assessment situation.

Also, what about the attribute and collection builder system? Surely that's not a breaking change, and is trivial to add in .NET 9? (At the cost of .NET 8 not getting this support.) Again, this applies to types where it makes sense - I'm not convinced `ColumnHeaderCollection` qualifies.

**Reply by TahirAhmadov** [reply]
For `ColumnHeaderCollection`, the manual (non-designer) control creation scenario I can imagine is:
```cs
var listView = new ListView
{
  Columns += [columnHeader1, columnHeader2, "Column 3"],
};
```
Here, we say that the imaginary future `+= []` syntax replaces the `= { }` collection initializer syntax, and is supported with new a attribute on the `ColumnHeaderCollection` class.

**Reply by CyrusNajmabadi** [reply]
> Do you mean the intention is to support something like [columnHeader1, columnHeader2, "Column 3"], etc.? I think this looks weird and should not be supported.

Yes.  it is fine writing that with a collection-initializer.  Importantly, it was a core part of hte design of collection-initializers to allow *exactly* this sort of pattern on a collection type.  Nothing has ever changed in that regard.  So APIs *have* been designed with this being supported and intentional.  It doesn't make sense to abandon the normal practices used by our first-party APIs that use exactly the patterns we designed the language to support.

**Reply by CyrusNajmabadi** [reply]
> I just cannot imagine

We have decades of experience seeing breaks of all sorts.  Making changes whne you have billions to trillions of lines out there is *risky*.  I have lost count at this point of hte number of times people thought a change would be totally fine, and it ends up taking something down.  

Pushing on people to take on risk because we regressed something is literally punishing the ecosystem for our mistakes.

**Reply by CyrusNajmabadi** [reply]
> Also, what about the attribute and collection builder system? Surely that's not a breaking change

First, realize that you're stating an implicit position that we shoudl start by having libraries have to do work to be functionality with collection exprs.  That goes explicitly against the goals of the feature, and the stated promises we've given to customers.  CollectionBuilder exists for the types that didn't work as collection *to begin with*.  It's not intended to be used to get types that are *completely normal collection types*, that already worked fine for 20+ years, to now work in the world of today.

Furthermore, this is regression on a supported scenario.  We started out with this *working*, and we *shipped* a real release that way.  Breaking it, and forcing the ecosystem to take lots of pain, just to make changes to make it work again benefits no one here.  It actively makes life painful for the api author and their customers, to workaround our mistake.

**Reply by CyrusNajmabadi** [reply]
> I'm not convinced ColumnHeaderCollection qualifies.

You're presuming all sorts of things about the ecosystem. For example, that people aren't stitching up these headers, and then using them in things like their own UI types.  Or that they're not making these collections in one place, but then imperatively processing them elsewhere to instantiate UI components.

These are apis that are decades old, with *tons* of code written by millions of users for tons of companies with tons of varying styles and approaches to do their work.  We cannot assume everyone uses them in precisely one fashion, and then go "this break is ok, because everyone will be doing it in this one way".  The real world needs of compat are broad and varied.  

We cannot assume people use designers.
We cannot assume people construct all values up front and immediately assign them.
We cannot assume people only use one of the Add methods.
We cannot assume people are not doing normal type checks on their instances.
etc. etc. etc. 

Think about just how far back winforms goes.  And how *heavily* it is used.  Indeed, it was precisely taht deep usage that has keep it something we continue to support to this day, precisely because it isn't niche.  It's a backbone system for tons of companies, and none of those companies are doing anything wrong in continuing to use thsi system we've espoused and maintained for them all this time.  Those customers are not lesser for doing this.  They're solving the needs of their business, and are just as relevant to those using different tech stacks.  Having them be able to benefit from the changes the lang makes is a good thing :)

**Reply by TahirAhmadov** [reply]
That's the thing - I don't think `ColumnHeaderCollection` is a collection to begin with. It's even less of a collection than other types where CollectionBuilder stuff was done in the API. It can't work currently at all because the ctor expects `ListView` parameter. 
I think the LDM and the notes are based on a type which is a fundamentally bad example for this conversation.

**Reply by CyrusNajmabadi** [reply]
Taht's probably true.  The point as looking at how several collection-esque types allow `.Add(string)` (which would be fine for a collection-initializer space).   We can pick some other type if that makes people more comfortable.  But it doesn't change the argument at all, since we're not saying we want this specific type to work, just types like it that would otherwise have worked when we shipped C#12 but now are broken.  That means supporting types that allowed for different overloads of 'Add' just as collection-initializers always allowed.

**Reply by TahirAhmadov** [reply]
It's not as much about being "comfortable", rather about being able to look at specific options and whether they help this or that type. I almost think the maintainers of WinForms, WPF, etc. should pitch in and tell us which types they think should be supported and how. What if all these "collection-esque" types turn out to be not really good candidates for C.E.s at all once you look at each one of them? Then it would be a shame that all this time was wasted discussing a non-problem.

Regarding interfaces, for what it's worth, some types got `IEnumerable<T>` implementation, like `MatchCollection`, and that didn't warrant this level of alarm about potential breaks. Again, input from library maintainers would be very helpful here.

**Reply by CyrusNajmabadi** [reply]
Put another way, we've never dissuaded anyone from this aspect of the language in all the time since we shipped this support.  Collection-initializers intentionally supported this pattern precisely because it was fine.  And we shipped C#12 respecting that.  The issue is dealing with a post-ship change that broke this support.

**Reply by TahirAhmadov** [reply]
I'd like to underscore the importance of concrete examples. In this case, a language decision was made, which potentially precludes some collection types from participating in a certain way. The way I see it, it's possible that either 1) the language change was in error and that these collections need to be supported, or 2) the language change was correct and these types are not really collections, and the fact that they are not supported being unexpected does not make it a change in the wrong direction. Sometimes unexpected changes are good.

**Reply by CyrusNajmabadi** [reply]
> the language change was correct and these types are not really collections,

The types are collections.  We made that very clear in the design of collection-initializers.  

> Sometimes unexpected changes are good.

These changes have already broken customers.  Customers who have collection types that work completely fine with collection-initializers, using the *exact pattern* we have documented and espoused since C#3.0  (nearly 17 years old).  In all that time we have never once said the pattern is wrong or undesirable.  Furthermore, when creating, documenting and discussing collection expressions with those same customers, we stated explicitly that it was intentional and a goal of the feature that those cases would move forward.  

It is expected that users who use collection-initializers be able to move them forward.  The only cases not supported are those we deliberately looked at and said we didnt' have an actual solution for (for example, providing a collection argument, or being multi-dimensional).  However, outside of the carved out cases (which are thinsg we *do* want to look into supporting for the future mind you, so that everyone can move forward), accidental blockages *especially* on shipped support is not desirable.

**Comment by iam3yal** [comment]
Just wondering whether the keyword `let` was discussed? as opposed to `allow` or even `allows`.

**Reply by colejohnson66** [reply]
I don't like that. `let` is commonly associated with binding (JS, Rust, F#, etc.), not constraints. Using it in a generic constraint would read weird: `where T : let ref struct`.

**Reply by jnm2** [reply]
`let T : ref struct` sounds a bit less awkward, but the decision was already made that having to look in two different clauses to understand `T` was not preferable.

**Reply by iam3yal** [reply]
@colejohnson66 I don't disagree but yes my thoughts were more aligned with the way @jnm2 expressed it. 

Not that it matters but personally I dislike `allows` and `allow` even more but I'm not opposed to it, I trust the LDT to make the right decision. :)

**Reply by 333fred** [reply]
We did not discuss `let`. I could see it working for the separate constraint scenario, but not for the inside `where` scenario, which we felt much more strongly about.

**Comment by jnm2** [comment]
For flow of allow/allows, I can think of two conceptualizations that make sense to me when reading aloud:

1. "where `T` allows ref struct": `T` is the subject
2. "where [there is a] T, allow ref struct": this is either imperative, or "[we/I] allow"

The first form would be read without a pause, and the second form would be read with a pause at the comma.

But now let's consider `where T : class`:

1. 'where T is [a] class": `T` is the subject
2. 'where [there is a] T, [it must be a] class": this is no longer imperative or using "we/I"
   - or, 'where [there is a] T, [require it to be a] class": this is either imperative or "[we/I require]"

The thing that feels the most streamlined and least contrived to me is strategy 1.

**Comment by HaloFour** [comment]
Probably important to look at the "anti-constraint" in the context of being combined with other constraints as well, as I can see that being a common use case (particularly with specific interfaces):

```cs
where T : allows ref struct, IDisposable
// or
where T : allows ref struct, struct, IDisposable
// or
where T : struct, allows ref struct, IDisposable
```

**Reply by TahirAhmadov** [reply]
I think this is important, actually. I think it should be:
```cs
where T : struct, IDisposable; allows ref struct
// or
where T : struct, IDisposable, allows ref struct
```

**Reply by HaloFour** [reply]
From the [proposed spec](https://github.com/dotnet/csharplang/pull/7977/files#diff-22d29a977b526e425bb1b0142fabacf08bb7bdeb888c622b8a7b5ffa15410d1eR157), it appears that this is the case and that `allows` clause must be the last clause.  I also think this makes sense as it could be easy to confuse `allows ref struct, IDisposable` as the anti-constraint also applying to `IDisposable`.

**Comment by huoyaoyuan** [comment]
As a non-fusional language speaker, `allow` vs `allows` is totally nothing to me. Personally I prefer `allow`.

**Reply by ufcpp** [reply]
Come to think of it, C# doesn't have a keyword with s of the third person singular, while Java has implements and throws.

**Comment by alrz** [comment]
>  anti-constraint 

It is true that this is an anti-constraint - allowing more types to be used rather than less, but it still limits the lifetime? I remember a proposal for lifetime annotations, isn't that going to happen? (even as a subset, but I think that could drive the syntax form)

> cannot participate in methods that take `IReadOnlyList<T>`, `IEnumerable<T>`

The proposal mentions this as a motivation but this feature would not enable that unless the method takes a generic parameter. Wouldn't it make sense to wait for extensions to implement interfaces on ref structs? As per the last discussions I think they would always require a generic parameter.

I'm looking out for `Span<Span<T>>` support, but I think none of these would help with that? I don't know if that's even possible though.

**Reply by 333fred** [reply]
>  I remember a proposal for lifetime annotations, isn't that going to happen?

No, we never planned to actually do that. It was a form of proof that our rules were sound.

**Comment by jubruckne** [comment]
I don’t like this syntax much. Why can we just have more flexible constraints like:
``` 
where T:struct or ref struct
```

and maybe in the future even like: 
```
where T:float or int, INumber<T>
where T:struct, Vector2 or Vector3
```

**Reply by colejohnson66** [reply]
Short answer: because the CIL/runtime doesn't support it.

Long answer: These types of constraints are easy for dynamic languages like TypeScript, but harder to compile for. The compiler would be required to look at the list of methods available on any of the targets, then find the union of them all. Only that union subset would be available to use. IMO, it's a bad idea, as methods with the same signature can have different semantics. Interfaces are supposed to solve that issue, as an interface contract ensures implementers have the same semantics.

Is it possible? Sure. But will it ever happen? Doubt it. Overloads and `where T:IFace` exist, and both are the recommended way to handle these situations.

**Comment by HaloFour** [comment]
Another possible "anti-constraint" that came up on Discord was one that, when combined with the `struct` constraint, would permit a nullable value type (e.g. `int?`) to be the generic type argument.  Seems the runtime already supports this via existing combinations of constraints but C# has never exposed it.

/cc @IS4Code

**Comment by Ghost4Man** [comment]
The "2023" in the title (`LDM Notes for February 26th, 2023`) is a typo, right?

**Reply by 333fred** [reply]
Yes, thanks.
